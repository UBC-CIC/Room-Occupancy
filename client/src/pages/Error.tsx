import * as React from "react";
import { Header } from "semantic-ui-react";
import "../App.css";
import { Button } from "semantic-ui-react";
import { Link } from "react-router-dom";

export interface IErrorProps {}

export function Error(props: IErrorProps) {
  return (
    <div className="container">
      <Header size="huge" color="blue" textAlign="center">
        <Header.Content>404!</Header.Content>
        <Header.Subheader color="blue">
          Oops! This page could not be found!
        </Header.Subheader>
        <Link to="/">
          <Button circular color="blue">
            Go to Homepage
          </Button>
        </Link>
      </Header>
    </div>
  );
}
