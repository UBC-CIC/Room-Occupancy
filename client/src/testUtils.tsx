import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import {
  MemoryRouter,
  MemoryRouterProps,
  BrowserRouter,
} from "react-router-dom";

// Implement with creation of providers
const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return <MemoryRouter>{children}</MemoryRouter>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

// Usage example:
// test("matches snapshot", () => {
//   // Replace 'Navbar' with your desired component and provide the props
//   testSnapshot(Navbar, { /* props for Navbar */ });
// });
const testComponentSnapshot = <P extends {}>(
  Component: React.ComponentType<P>,
  props: P,
  routerProps: MemoryRouterProps = {}
) => {
  const { asFragment } = render(
    <MemoryRouter {...routerProps}>
      <Component {...props} />
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
};

export * from "@testing-library/react";
export { customRender, testComponentSnapshot };
