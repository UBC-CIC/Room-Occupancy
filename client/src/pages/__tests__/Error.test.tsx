import React from "react";
import { render, screen, act } from "@testing-library/react";
import { Error } from "../Error";
import { MemoryRouter } from "react-router-dom";
import "@testing-library/jest-dom";
import { testComponentSnapshot } from "../../testUtils";

test("renders 404 header", () => {
  act(() => {
    render(
      <MemoryRouter>
        <Error />
      </MemoryRouter>
    );
  });
  const headerElement = screen.getByText("404!");
  expect(headerElement).toBeInTheDocument();
});

test("should match snapshopt", () => {
  testComponentSnapshot(Error, {});
});

test("renders error message", () => {
  act(() => {
    render(
      <MemoryRouter>
        <Error />
      </MemoryRouter>
    );
  });
  const errorMessageElement = screen.getByText(
    "Oops! This page could not be found!"
  );
  expect(errorMessageElement).toBeInTheDocument();
});

test("renders 'Go to Homepage' button", () => {
  act(() => {
    render(
      <MemoryRouter>
        <Error />
      </MemoryRouter>
    );
  });
  const buttonElement = screen.getByText("Go to Homepage");
  expect(buttonElement).toBeInTheDocument();
});

test("navigates to homepage on button click", () => {
  const { container } = render(
    <MemoryRouter initialEntries={["/colins-route"]}>
      <Error />
    </MemoryRouter>
  );
  const buttonElement = screen.getByText("Go to Homepage");
  expect(buttonElement).toBeInTheDocument();

  act(() => {
    buttonElement.click();
  });

  const currentLocation = container.querySelector("a")?.getAttribute("href");

  expect(currentLocation).toBe("/");
});
