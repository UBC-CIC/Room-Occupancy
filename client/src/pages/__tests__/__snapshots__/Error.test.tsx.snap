// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshopt 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="ui blue huge center aligned header"
    >
      <div
        class="content"
      >
        404!
      </div>
      <div
        class="sub header"
        color="blue"
      >
        Oops! This page could not be found!
      </div>
      <a
        href="/"
      >
        <button
          class="ui blue circular button"
        >
          Go to Homepage
        </button>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
